### ISSUE TITLE
Font scaling causes clipping at high resolution

### ISSUE THREAD

**Issue by morganwillcock** [adventuregamestudio/ags#447]
As an example, current project is 1600x1200 and fonts are imported at size 36. Whichever font I use, the bottom of some characters (p, y, g, etc.) are getting cut off. Changing renderer makes no difference, but I can get complete characters by specifying a negative vertical offset for the font.  Changing the "Fonts designed for high resolution" setting doesn't improve the situation or make it worse.

To summarise:

- When VeritcalOffset is set to -2 the characters appear complete
- When VerticalOffset is 0, the lower parts of some characters are not visible.

**Comment by ivan-mogilko** [comment]
Question: does same happen if font is imported with lower size?

This is probably same issue that was reported couple of years ago (on forums), and this is actually what VerticalOfsset was added for.
The reason is most probably that we are using old alfont library, that is in turn based on very old version of FreeFont. In the past I did experiments and found out that upgrading to latest FreeFont lib solves such issues.

(But in turn introduces new ones, stopping displaying some of the fonts in old games correctly. But these are apparently broken in design)

**Comment by ivan-mogilko** [comment]
For the reference, this is a branch I was experimenting with: https://github.com/ivan-mogilko/ags-refactoring/commits/libs--alfont

**Comment by morganwillcock** [comment]
> Question: does same happen if font is imported with lower size?

No, importing at a smaller size doesn't have the same problem (just re-tested by re-importing at size 10)

> For the reference, this is a branch I was experimenting with: https://github.com/ivan-mogilko/ags-refactoring/commits/libs--alfont

I'm not sure how far @AlanDrake got with stripping out old features, but if dropping all backwards compatibility on his branch, perhaps we can just make these sort of changes there (so this sort of thing, plus remove all the legacy audio support, etc).

**Comment by ivan-mogilko** [comment]
In theory, the broken fonts could be replaced by placing font files in game directories, so I believe this may be even done in this branch.
I now recall thinking doing that after 3.4.1 release. Did not do that earlier, being afraid to break something.

**Comment by ivan-mogilko** [comment]
Regarding version with removed legacy stuff... Thing is that we continue to add stuff to master. and personally I would not like to develop in two versions at the same time, so it's either this or that.

**Comment by AlanDrake** [comment]
If you wish I can make time to update the clean-up branch this weekend. Then we can merge it on master to avoid further derailment.
I think the previous plan of keeping one branch only for hotfixes and this one for new development is still a solid one.

**Comment by ivan-mogilko** [comment]
@AlanDrake
> If you wish I can make time to update the clean-up branch this weekend.

If you could make a clean rebase on top of master that would be awesome. Iirc it contains only parts of code removal? If not, oh well.

@AlanDrake 
> I think the previous plan of keeping one branch only for hotfixes and this one for new development

It was, but these changes were never merged, and as much time passed some people asked for certain new features, so I did another wip version called 3.4.2. If the cleanup happens, then either a) we need to tell them that the version breaks compatibility completely and call it 4.0.0, or b) I'd have to split another branch out for finishing 3.4.2.
I fear that it is going to be latter, because there is another problem:  if we actually do that we would like to make sure we remove all of legacy support at least from the editor/script API side, and so cleaning up legacy may take a while, much longer to wait for an actual release. (maybe not, but who can tell)

In any way this change won't be organizationally simple. Firstly some people keep using old API because they don't like new one for some reasons; in which case we will probably have to work out good alternative. Secondly because there are users who already use latest master branch to run older games, that is for instance they already having their savedgames in newest format ([there was already an issue because of this](https://github.com/adventuregamestudio/ags/pull/438#issuecomment-379563790)).
Third, I've already expressed this worries before, but if we make this split, this would mean that code refactoring and any program optimisation would only benefit new version, but not the one running old games (and that is extensively used to run old games on Linux, for example).

(Of course there is a scummvm port that @morganwillcock was working on lately, so it could provide alternative when finished, but idk how long that would take)

And still there issue imho is a lack of set convention (rules) for the further development. In the past it was clearly told that the project's aim was (among other things) to support as much old games as possible. So far it was only you who expressed the wish to remove certain code and functionality, but it's not enough for the open source project. We will need to write something in the readme too, explaining what users/developers should expect, how people should deal if they want to continue develop a version which supports running old games, what kind of things are allowed or permitted in the main branch etc.

Last but not least, I hope everyone realizes, that if we start this cleaned branch, there is a good chance that we won't be able to release a new stable version for a while. I mean, it's not necessary be like that, but possible. And to tell the actual truth, my biggest concern if that happens is how many contributors will be around to help with fixing possible issues. Because people ask for new stuff all the time, and sometimes want to get it asap for their game. If it's something small, I will have very hard time explaining why we cannot give them a new version.

**Comment by ivan-mogilko** [comment]
> I'm not saying importing older formats shouldn't be a thing, but it'd be better to write that code in a standalone module. I don't know how the format is going to evolve, but it'd be nice if it could evolve.

So, would it be right to say, that the idea is to first clean everything up, then return useful bits back in such way that won't restrict new code? I think this should be made clear for the future reference.

**Comment by morganwillcock** [comment]
> So far it was only you who expressed the wish to remove certain code and functionality, but it's not enough for the open source project.

> (Of course there is a scummvm port that @morganwillcock was working on lately, so it could provide alternative when finished, but idk how long that would take)

The main reason I want to get this port usable is so that everything but current features could be dropped if needed (whether that be for a cleaned up AGS or an entire replacement), so I guess you can count me as +1 to only keep the essentials and remove support for all of the edge cases (and what exists now becomes bug-fix only).

I should probably open a new issue with what I think should be done, because all of my previous old notes were on the Team City server, plus I've had a chance to look around more since then...

> The cleanup branch was already in a relatively stable condition back then, the only problem is older version formats and the like, which is quite a problem in the current game templates.

The next thing I'm going to look at are the templates and a test game, so hopefully this should get resolved.

**Comment by ericoporto** [comment]
Just to add to the list of failed experiments xD, https://github.com/ericoporto/ags/commits/upgrade-freetype

I looked into pushing to the latest FreeType 2.10.4, it works alright with most fonts, but you are bound to find a game where the fonts just won't look right. This is the version SDL_ttf currently uses though, and I imagine switching Alfont with SDL_ttf is in the realm of possibilities (they have sorta similir APIs), but since it will probably modify things a bit, it should be a ags4 development.

**Comment by rofl0r** [comment]
> I looked into pushing to the latest FreeType 2.10.4

FYI, the only reason we added freetype source code to the repo instead of using system-provided freetype (which would be the much better option) is that the ancient version is the one used by CJ's binary builds and the only one that works correctly with lucas-arts font, which many old games use.

**Comment by ericoporto** [comment]
> lucas-arts font, which many old games use.

Yeah, exactly this one from maniac mansion games, it breaks the outline!

**Comment by ericoporto** [comment]
got things almost working with sdl_ttf: https://github.com/ericoporto/ags/tree/sdl_ttf

missing figuring out how to properly blend SDL_Surface to Allegro bitmaps, I am messing something with alphas so I am getting a black box around my TTF texts (needs a game with ttf fonts to test, since the wfn renderer is ags internal and not in alfont).
And also need to fix the vcxproj files and add stl_ttf to props.

**Comment by ivan-mogilko** [comment]
In regards to the original problem -

> As an example, current project is 1600x1200 and fonts are imported at size 36. Whichever font I use, the bottom of some characters (p, y, g, etc.) are getting cut off. Changing renderer makes no difference, but I can get complete characters by specifying a negative vertical offset for the font

Working on an unrelated task I maybe could experimentally find a fix for this, by changing `alfont_text_height` from
`return f->face_h;`
to
`f->face_ascender + abs(f->face->size->metrics.descender >> 6);`

(note that in current code this is actually equivalent to `f->face_h + abs(f->face->size->metrics.descender >> 6);`, because we store same value in both face_h and face_ascender).

With the above change I was able to fix both the problem I was working on, and also large fonts getting centered incorrectly and thus having their bottoms cut off.

**Comment by ivan-mogilko** [comment]
I will link my comment from another ticket, where I think I found the correct way to apply backward compatible alfont hacks:
https://github.com/adventuregamestudio/ags/issues/888#issuecomment-899857897

I'd test the game with these changes to see if they fix this bug in hi-res game too.

**Comment by ivan-mogilko** [comment]
The patched engine code for the test may be built from here: https://github.com/ivan-mogilko/ags-refactoring/tree/alfont-fix-test

**Comment by ivan-mogilko** [comment]
Pushed a change to alfont, fixing another issue: 41631638eca0bfab4306c631f5d0d0780889c3ad

I don't know if it will fix this too, or it's something different. Need to retest.

**Comment by ivan-mogilko** [comment]
@morganwillcock I was able to produce some error by increasing the font size; at some point the drawing surface created for the speech etc becomes not enough to fit the text due to the font height mistake, and the last line of the wrapped text becomes clipped. Don't know if this is also related to the case you experienced.

I pushed few more commits: 8a735b42d8f6079edbdbd5f1968a456decf77487 47001a398a1bc4ad022645ad21e1c891a3757167 3a9b8f199e09b9919a201c1d2fa486a79761b339
These make size calculation depend on the real TTF font size.

To keep text wrapping backward compatible I made linespacing to be assigned a "formal" font height by default, in case linespacing was not specified explicitly in the game properties. This keeps line wrapping same. The total height of all wrapped lines may be slightly different, but I hope that this won't cause much problems. If later we find it does, there's an opportunity to add a switch depending on a game version.

Similarily, later we could add a switch for linespacing to not depend on "formal" font height but on real font height - if wanted. But with linespacing the situation is simplier, as it may be assigned exact value in properties.

**Comment by ivan-mogilko** [comment]
@morganwillcock 

An update on this -
1) I programmed engine to try account for the real pixel height of a font whenever it draws speech and messages, so in that case it should no longer be cut at the bottom. This does not help if you put text on GUI and resize that GUI to the font's import size value. For that there are other solutions -
2) As of 3.6.0.11 there are now two new settings, one called "TTF font adjustment", it's a per-font property, that lets disable AGS historical hack where it would resize font's ascender, which could result in shifting the text output up or down unexpectedly (there's also a property of same name in General Settings that lets to choose a project's default value for new fonts).
3) Also there's a global "TTF fonts height used in the game logic" which lets to enforce use of real pixel height of a TTF in the text placement and GetTextHeight return values.

I think that the combination of these should fix this issue.

**Comment by ivan-mogilko** [comment]
Unfortunately op did not reply since, but I believe this should be fixed in 3.6.0 now.
